python/summarize_sgfs.py

```diff
@@ -0,0 +1,295 @@
+
+  def _add_new_games_to_result_dict(self):
+    """add all sgfs files first"""
+    idx = 1
+    for sgfs in self._new_sgfs_files:
+      self._add_one_sgfs_file_to_result(sgfs)
+      if (idx % 10 == 0):
+        print(f"Addedd {idx} files out of {len(self._new_sgfs_files)} sgfs files.")
+      idx += 1
+    print(f"We have added additional {len(self._new_sgfs_files)} sgfs files into the results.")
+
+    idx = 1
+    for sgf in self._new_sgf_files:
+      self._add_one_sgf_file_to_result(sgf)
+      if (idx % 10 == 0):
+        print(f"Added {idx} files out of {len(self._new_sgf_files)} sgf files.")
+      idx += 1
+    print(f"We have added additional {len(self._new_sgf_files)} sgf files into the results.")
+
+  def _add_one_sgfs_file_to_result(self, sgfs_file_name):
+    """Add a single sgfs file. Each row of such sgf file contain a single game as a sgf string"""
+    if not os.path.exists(sgfs_file_name):
+      print(f"There is no SGFs file as {sgfs_file_name}")
+      return None
+
+    with open(sgfs_file_name, "rb") as f:
+      sgfs_strings = f.readlines()
+
+    for sgf in sgfs_strings:
+      self._add_a_single_sgf_string(sgf)
+
+  def _add_one_sgf_file_to_result(self, sgf_file_name):
+    """Add a single sgf file."""
+    if not os.path.exists(sgf_file_name):
+      print(f"There is no SGF file as {sgf_file_name}")
+      return None
+
+    with open(sgf_file_name, "rb") as f:
+      sgf = f.read()
+
+    self._add_a_single_sgf_string(sgf)
+
+  def _add_a_single_sgf_string(self, sgf_string):
+    """add a single game in a sgf string save the results in self.results."""
+    game = sgf.Sgf_game.from_bytes(sgf_string)
+    winner = game.get_winner()
+    pla_black = game.get_player_name('b')
+    pla_white = game.get_player_name('w')
+    if (game.get_handicap() is not None) or game.get_komi() < 5.5 or game.get_komi() > 7.5:
+      self._warning = True
+
+    if (pla_black, pla_white) not in self.results:
+      self.results[(pla_black, pla_white)] = Record()
+
+    if (winner == 'b'):
+      self.results[(pla_black, pla_white)].win += 1
+    elif (winner == 'w'):
+      self.results[(pla_black, pla_white)].lost += 1
+    else:
+      self.results[(pla_black, pla_white)].draw += 1
+
+  def _estimate_elo(self) -> elo.EloInfo:
+    """Estimate and print elo values. This function must be called after add all the sgfs/sgf files"""
+    pla_names = set(itertools.chain(*(name_pair for name_pair in self.results.keys())))
+    data = []
+    for pla_Balck in pla_names:
+      for pla_White in pla_names:
+        if (pla_Balck == pla_White):
+          continue
+        else:
+          record = self.results[(pla_Balck, pla_White)]
+          total = record.win + record.lost + record.draw
+          win = record.win + 0.5 * record.draw
+          winrate = win / total
+          data.extend(elo.likelihood_of_games(pla_Balck, pla_White, total, winrate, False))
+    data.extend(elo.make_center_elos_prior(list(pla_names), self._elo_prior))
```

Comment:
Sorry, I probably wasn't clear. :)

I didn't mean that you should be able to set the center of the elo prior. I meant that you should have an argument that controls the weight of a prior that for each player, pulls that player's Elo towards 0.

So, loop over all players and extend data with `elo.make_single_player_prior` with `num_games` based on the user's input. This will provide actual regularization of the Elos for players with too little data, not just center them.

Reply:
Below is the changes based my understanding on this comment. Please see if this change is right. This might be the one that I am not 100% sure (but 80% sure) on comment. The propose of the prior is to stabilize small number of games having big impact, and we still need to center the Elo at 0. So the changes will be,

   ```
 for pla in pla_names:      
      data.extend(elo.make_single_player_prior(pla, self._elo_prior,0))      
    data.extend(elo.make_center_elos_prior(list(pla_names), 0))
```

Reply:
This seems fine to me. Adding a 0 prior to all players will make it so that make_center_elos_prior is redundant, because if all players individually have a 0 prior, the collective error will also be minimized when their average is 0. But it also does no particular harm to have make_center_elos_prior as well.

Reply:
I agreed that is redundant. Let's remove `data.extend(elo.make_center_elos_prior(list(pla_names), 0))`.